### ISSUE TITLE
Support envFile in launch configuration

### ISSUE THREAD

**Issue by christinachimi** [Dart-Code/Dart-Code#3189]
**Is your feature request related to a problem? Please describe.**
My project uses a shared launch.json to facilitate development. When launching to an emulator, this is perfect. But when using a real device, I need to point my project to my local ip. Currently the way we do this is by editing the dart-define in our launch.json, then removing the edits before pushing to version control.

**Describe the solution you'd like**
I'd like to be able to set an env like `LOCAL_IP` and reference that in a shared launch configuration. Currently the only way I can see to do this is by setting `LOCAL_IP` as an `env` in my launch config, which defeats the purpose. There are a few ways to attack this but the most straightforward seems to be supporting `envFile`.

**Describe alternatives you've considered**
I have looked to see if VS Code supports user-specific launch configs (not ideal but would be fine as our configs rarely change) but I don't think that they do. In theory you should have access to environment variables within the launch config but in practice this doesn't seem to be the case.

**Additional context**
Example launch config:

```
{
  "name": "Flutter: Launch Main App (android)",
  "request": "launch",
  "type": "dart",
  "program": "lib/main.dart",
  "args": [
    "--dart-define=my.api=http://10.0.2.2:5378/api"
  ]
}
```

**Comment by DanTup** [comment]
>  In theory you should have access to environment variables within the launch config but in practice this doesn't seem to be the case.

That should actually work fine. For example:

```js
{
	"version": "0.2.0",
	"configurations": [
		{
			"name": "flutter_application_4",
			"request": "launch",
			"type": "dart",
			"args": [
				"--foo=${env:USER}"
			]
		}
	]
}
```

Running this, I see this in the logs:

```
Spawning flutter with args ["run","--machine", /*SNIP*/ "--foo=danny","-v"
```

Does this address your need?

**Comment by christinachimi** [comment]
How are you setting USER? I am not seeing my env variables being made available to the launch config. I'm on a mac, maybe that's relevant. I don't see any logs about spawning Flutter so I can't tell if this is the environment or an interpolation issue.

**Comment by DanTup** [comment]
`$USER` was just an existing env variable I have set in my system (I ran `printenv` and just picked one). I'm also on a Mac.

There's no completion for the env variables, but it's documented here:

https://code.visualstudio.com/docs/editor/variables-reference#_environment-variables

**Comment by jamesblasco** [comment]
I would love to have support for envFile in vscode for dart. It would be the perfect way to set up secret environment variables instead of adding them manually like "--dart-define= ...".

Right now it is possible to do this, but it is not very secret. Would love to load this env variables from a file
{
	"version": "0.2.0",
	"configurations": [
		{
			"name": "flutter_application_4",
			"request": "launch",
			"type": "dart",
                        "env" {
                             "SECRET_TOKEN": "___"
                        }
		}
	]
}

**Comment by DanTup** [comment]
> Right now it is possible to do this, but it is not very secret

Can you elaborate on what you're trying to do a little more? Wouldn't having another file with the environment variables just as "not very secret" as having them here? Or is the issue that you want to commit the `launch.json` but would `gitignore` the env file?

You could also set the values as _real_ environment variables on your system, or in the `dart.env` setting (which can be used at the user level, or project/folder level) today without needing any changes.

**Comment by pdmshrestha** [comment]
In Node.js we can load environment variables from external file https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_load-environment-variables-from-external-file-node. It would be nice if we can have similar way in dart too.

**Comment by DanTup** [comment]
@pdmshrestha see my comments above - I'd like to full understand the use case and problems with using the other mechanisms before doing anything here - unlike Node, Dart doesn't have anything built-in to load from .env files so it'd mean parsing the file (and understanding the exact syntax that is normally supported).

**Comment by pdmshrestha** [comment]
> > Right now it is possible to do this, but it is not very secret
> 
> Can you elaborate on what you're trying to do a little more? Wouldn't having another file with the environment variables just as "not very secret" as having them here? Or is the issue that you want to commit the `launch.json` but would `gitignore` the env file?
> 
> You could also set the values as _real_ environment variables on your system, or in the `dart.env` setting (which can be used at the user level, or project/folder level) today without needing any changes.

@

> > Right now it is possible to do this, but it is not very secret
> 
> Can you elaborate on what you're trying to do a little more? Wouldn't having another file with the environment variables just as "not very secret" as having them here? Or is the issue that you want to commit the `launch.json` but would `gitignore` the env file?
> 
> You could also set the values as _real_ environment variables on your system, or in the `dart.env` setting (which can be used at the user level, or project/folder level) today without needing any changes.

@DanTup  .env file would be git ignored. launch.json file will have various other configurations which we like to have in repository.

**Comment by pdmshrestha** [comment]
@DanTup  Yes. I can add environment variables to `launch.json` and git ignore it. But I would like to keep `launch.json` on git and only hide environment variables. For example my current configuration example is
```
// launch.json
{
"version": "0.2.0",
"configurations": [
{
      "name": "Development",
      "request": "launch",
      "type": "dart",
      "program": "./lib/main.dart",
      "showMemoryUsage": true,
      "env": {
        "CLIENT_SECRET": "*********",
        "API_URL": "https://1257fa8c220d.eu.ngrok.io/api/mobile-v2",
        "CLIENT_ID": 2,
        "BUSGNAGS_KEY": "*********"
      },
      "args": [
        "--dart-define=CLIENT_SECRET=${CLIENT_SECRET}",
       "--dart-define=CLIENT_ID=${CLIENT_ID}",
        "--dart-define=API_URL=${API_URL}",
       "--dart-define=BUSGNAGS_KEY=${BUSGNAGS_KEY}",
        "--flavor",
        "dev"
      ]
    },
//.... more configurations
]
}
```
The problem here is that my client environment variables are exposed. Which are different for teammates. Better solution would be provide envFile like in node.js app and use those variables on launch.json
```
// launch.json
{
"version": "0.2.0",
"configurations": [
{
      "name": "Development",
      "request": "launch",
      "type": "dart",
      "program": "./lib/main.dart",
      "showMemoryUsage": true,
      "envFile": "dev.env",
      "args": [
        "--dart-define=CLIENT_SECRET=${CLIENT_SECRET}",
       "--dart-define=CLIENT_ID=${CLIENT_ID}",
        "--dart-define=API_URL=${API_URL}",
       "--dart-define=BUSGNAGS_KEY=${BUSGNAGS_KEY}",
        "--flavor",
        "dev"
      ]
    },
//.... more configurations
]
}
```
dev.env
```
CLIENT_SECRET=*********
CLIENT_ID=2
API_URL=https://1257fa8c220d.eu.ngrok.io/api/mobile-v2"
BUSGNAGS_KEY=**********

```
With this setup `dev.env` could be git ignored. Another plus point for this is that we can use this same env file for other app for example if we using different tools for web development (react, vuejs)

**Comment by DanTup** [comment]
If you're not currently committing `.vscode/settings.json` to Git, you could put the env variables in there (in the `dart.env` setting) which will work today - although it doesn't allow sharing with non-Dart processes.

It does sound like there's some value in supporting something like this though - please add 👍 's to the original comment at the top to help gauge interest. Thanks!

**Comment by christinachimi** [comment]
@DanTup I came back to try this out based on your comments. I did find that I was able to get `${env:LOCAL_IP}` to work, but I think the issue that hung me up previously is that is unclear how to "source" these variables for VS Code (so in my testing it seemed like the variable wasn't available). Based on my experiments, you have to completely shut down VS Code (not just the workspace) in order for changes to propagate. For my use case this is no good so it would be another nice reason to have an envFile option here. Just wanted to provide that perspective, thanks for giving this consideration!

**Comment by DanTup** [comment]
@christinachimi the substitution in launch.json is performed by VS Code and is done based on the environment variables available to its process. If you modify them, I would expect that you'd need to restart VS Code to pick up the changes, though from the examples above it seemed like these things would be fairly static - is that not the case?

**Comment by christinachimi** [comment]
@DanTup just coming back to say that I think you're right, this is sufficient for our use case. I think you can probably close this unless you have other reasons to want to do this. Thanks for the discussion!

**Comment by DanTup** [comment]
@christinachimi thanks for the ping. I'll go ahead and close this if it's not pressing then - I do think there are better ways to do it (although VS Code's lack of local settings certainly make it less trivial than it could be) and with the debug adapter moving to the SDK, this would ultimately also need to be moved there (and supporting it there could be a harder sell).

I'm open to revisiting this is there are enough compelling cases (so future readers can add 👍 s and comments). Thanks!

**Comment by pseudo-su** [comment]
Ok, after a bit more time with it I realised this won't actually work because I think both `env` and `envfile` configuration in `tasks.json` are resolved after the `${env:THING}` resolvers.

For anyone who stumbles upon this in the future trying to get `envFile` to work:

https://github.com/microsoft/vscode/issues/89825#issuecomment-586920401

**Comment by DanTup** [comment]
> I would also typically have changes to `.env.local` gitignored because there's lots of changes that each person might want/need to make to their local dev version while developing

Understood. I think the real fix for this is for VS Code to support multiple local files (so one can be source-controlled, and one ignored). That's been a highly requested feature forever, but not implemented so far.

Another possibility is to put your launch configs in a `.code-workspace` file, with a a `.sample.code-workspace` in your repo, and the used `.code-workspace` file gitignored. That way, a new dev would copy the `.sample` one and remove `.sample` from the filename, then customise that file.

> I think both `env` and `envfile` configuration in `tasks.json` are resolved after the `${env:THING}` resolvers.

I'm not sure I fully understand this (although the discussion here was also about launch configurations, tasks are a little different again).

**Comment by jjangga0214** [comment]
For folks who need `envFile`, there's an alternative [`flutter_config`](https://pub.dev/packages/flutter_config), which enables you to use env vars not only in .dart source code but also AndroidMenifest.xml and Info.plist (ios).

**launch.json**
```json
  {
      "name": "flutter(mainapp)",
      "request": "launch",
      "type": "dart",
      "flutterMode": "debug",
      "env": {
        "ENVFILE": "${workspaceRoot}/frontends/mainapp/.env.generated" // set as you want.
      }
    }
```

**Comment by TimWhiting** [comment]
I would like this for a dart server I'm developing where I would like to have a preprod/prod .env files, and use them as `Platform.environment` variables not as dart-defines (though I'll use that solution for now, I'd rather have them as environment variables so I can just set them on Google Cloud as environment variables to my docker container). I'm committing the `launch.json` to share with the team but I don't want the environment secrets (DB username/password) leaking. With this being a common property among many debuggers I don't see why dart should not support it. Again the main use would be for servers and `Platform.environment` not dart defines.

**Comment by DanTup** [comment]
> With this being a common property among many debuggers I don't see why dart should not support it.

As I understand it, the debuggers that support this are those where the runtime already supports it (or the language has some framework for it). I don't believe any debuggers are actually parsing this file themselves.

If what you want is some file to put env variables in that you can read in code, there are some existing pub packages like https://pub.dev/packages/envify that will do this, but you'll need to access the values using their classes (eg. `Env.foo`) instead of `Platform.environment` (and I suspect most people are wrapping them up in a class for stronger typing anyway).

**Comment by DanTup** [comment]
FWIW in #4260 I was told about https://github.com/flutter/flutter/pull/108098 which adds support for reading dart-defines from a file. I don't think it's shipped in a stable Flutter yet, but when it does I suspect this will handle many of the use cases above without the need of an additional package.

**Comment by DanTup** [comment]
`--dart-define-from-file=` shipped in the revent Flutter 3.7 release. There's some info on using this here:

https://medium.com/@tatsu.ukraine/flutter-3-7-and-a-new-way-of-defining-compile-time-variables-f63db8a4f6e2
